org.eclipse.xtext.common.types.ui/src/org/eclipse/xtext/common/types/ui/notification/DeltaConverter.java

```diff
@@ -187,21 +189,46 @@ protected void convertChangedCompilationUnit(IJavaElementDelta delta, List<IReso
 		
 		ICompilationUnit compilationUnit = (ICompilationUnit) delta.getElement();
 		try {
-			for (IType type : compilationUnit.getTypes()) {
-				if (isDerived(type))
-					return;
+
+			boolean hasDerivedType = doForeachCompilationUnitChildType(compilationUnit, true, type -> {
```

Comment:
Hmm I'm not terribly convinced yet. The common case would be, that we do not have 100s and 1000s of CU. Also I'd expect that the CU is accessed in subsequent operations again, e.g. during the Xtext build. Eager removal from the caches does not seem to be the right thing to do here. Should this be guarded against the number of CU in the delta? Or against the size of the existing cache.
A system property could mitigate these concerns. If the property was set, we'd do the eager cache cleanup. Otherwise not.

What do you think?

Reply:
I've been making some tests and it's quite complex to understand and as you say, not beneficial in all the cases

In a project with 2 java files, one referencing another, JDT does not put/remove anything in the cache, but is linking them appropriately
When a file with a reference to another is open, both end up in the cache, and after closing it, the referenced java is still in there (in can get removed if it's open and then closed)

In another test with an xtend file (GenTestXtend0) referencing a java class (GenTestJava0_0) is more complex, because they are added in different cases:
* GenTestXtend0.java: when the xtend file is compiled to java and is gets added by DeltaConverter
* GenTestJava0_0: when it appears and is detected by DeltaConverter (in my case, generated by a main file) and also when building the xtend scope of GenTestXtend0 (when saving xtend, clean, etc.)

The fix solves GenTestJava0_0 being added by DeltaConverter, so it would use less memory if it's not referenced by an xtext model like in my case, but it will be added twice if it's referenced.

I've made a test with 1000 xtend files referencing 10 unique java files with 200 variables and 10 common java files referenced by all the xtend files.

With and without the fix, the max memory used (1.4Gb) and the total time seems to be the same (51s). The end memory is 370Mb (GenTestJava* are not cleared from cache and GenTestXtend*.java does not occupy a lot)

The same test but without xtend files is different
Max memory is similar with or without the fix: 400-500Mb
End memory is 300Mb without the fix and 150Mb with the fix (GenTestJava* are cleared from cache)

Anyway, maybe the root cause is using CompilationUnit + JavaModelCache to obtain Java type info? 
Somehow JDT does not use it to do a full clean of a java project, but i don't know how